external/kafka/src/main/scala/org/apache/gearpump/streaming/kafka/KafkaSource.scala

```diff
@@ -36,62 +37,105 @@ object KafkaSource {
+  extends TimeReplayableSource {
   import org.apache.gearpump.streaming.kafka.KafkaSource._
 
-  private[kafka] def this(appName: String, config: KafkaConfig, topicAndPartitions: Array[TopicAndPartition], messageDecoder: MessageDecoder) =
-    this(FetchThread(topicAndPartitions, config), messageDecoder,
-      topicAndPartitions.map(tp => tp -> KafkaOffsetManager(appName, config, tp)).toMap)
-
-  private[kafka] def this(appName: String, config: KafkaConfig, grouper: KafkaGrouper, messageDecoder: MessageDecoder) =
-    this(appName, config, KafkaUtil.getTopicAndPartitions(KafkaUtil.connectZookeeper(config)(),
-      grouper, config.getConsumerTopics), messageDecoder)
+  private lazy val messageDecoder = config.getMessageDecoder
+  private lazy val timestampFilter = config.getTimeStampFilter
+  private lazy val batchSize = config.getConsumerEmitBatchSize
+  private var startTime: Option[TimeStamp] = None
 
-  def this(appName : String, taskId: TaskId, taskParallelism:Int, config: KafkaConfig, messageDecoder: MessageDecoder) =
-    this(appName: String, config, config.getGrouperFactory.getKafkaGrouper(taskId, taskParallelism), messageDecoder)
+  def this(config: KafkaConfig) = {
+    this(config, None, Map.empty[TopicAndPartition, KafkaOffsetManager])
+  }
 
-  private var startTime: TimeStamp = 0L
 
   LOG.debug(s"assigned ${offsetManagers.keySet}")
 
-  /**
-   * fetchThread will start from beginning (the earliest offset)
-   * on start offset not set
-   */
-  def startFromBeginning(): Unit = {
-    fetchThread.setDaemon(true)
-    fetchThread.start()
-  }
 
-  override def setStartTime(startTime: TimeStamp): Unit = {
+  private[kafka] def setStartTime(startTime: Option[TimeStamp]): Unit = {
     this.startTime = startTime
-    offsetManagers.foreach { case (tp, offsetManager) =>
-      offsetManager.resolveOffset(this.startTime) match {
-        case Success(offset) =>
-          LOG.debug(s"set start offset to $offset for $tp")
-          fetchThread.setStartOffset(tp, offset)
-        case Failure(StorageEmpty) =>
-          LOG.debug(s"no previous TimeStamp stored")
-        case Failure(e) => throw e
+    fetchThread.foreach { fetch =>
+      this.startTime.foreach { time =>
+        offsetManagers.foreach { case (tp, offsetManager) =>
+          offsetManager.resolveOffset(time) match {
+            case Success(offset) =>
+              LOG.debug(s"set start offset to $offset for $tp")
+              fetch.setStartOffset(tp, offset)
+            case Failure(StorageEmpty) =>
+              LOG.debug(s"no previous TimeStamp stored")
+            case Failure(e) => throw e
+          }
+        }
       }
+      fetch.setDaemon(true)
+      fetch.start()
+    }
+  }
+
+  override def open(context: TaskContext, startTime: Option[TimeStamp]): Unit = {
+    import context.{appName, parallelism, taskId}
+
+    val grouper = config.getGrouperFactory.getKafkaGrouper(taskId, parallelism)
```

Comment:
`KafkaConfig` is to make kafka related configs easy to use. `kafkaConfig.getGrouperFactory` instead of `userConfig.get("kafka.grouper.factory.class").getOrElse(classOf[KafkaDefaultGrouperFactory].getName)`